ddlitlab2024/ml/training/train.py

```diff
@@ -1,130 +1,129 @@
+    train_denoising_timesteps = 1000
+
+    # Load the dataset
+    logger.info("Create dataset objects")
+    dataset = DDLITLab2024Dataset()
+    num_workers = 5
+    dataloader = DataLoader(
+        dataset,
+        batch_size=batch_size,
+        shuffle=True,
+        collate_fn=DDLITLab2024Dataset.collate_fn,
+        persistent_workers=num_workers > 1,
+        # prefetch_factor=10 * num_workers,
+        num_workers=num_workers,
+        worker_init_fn=worker_init_fn,
     )
 
-    num_samples = real_trajectories.size(0)
-
-    # Subplot each joint, showing the first n batches
-    n = 1
-    plt.figure(figsize=(12, 6))
-    for i in range(trajectory_dim):
-        plt.subplot(3, 4, i + 1)
-        plt.plot(real_trajectories[:n, :, i].T)
-        plt.title(f"Joint {data.columns[i]}")
-    plt.suptitle("LKnee Trajectories")
-    plt.show()
+    # Get some samples to estimate the mean and std
+    logger.info("Estimating normalization parameters")
+    num_normalization_samples = 10
+    random_indices = np.random.randint(0, len(dataset), (num_normalization_samples,))
+    normalization_samples = torch.cat([dataset[i].joint_command_history for i in tqdm(random_indices)], dim=0)
+    normalizer = Normalizer.fit(normalization_samples.to(device))
 
     # Initialize the Transformer model and optimizer, and move model to device
-    model = End2EndDiffusionTransformer(
-        num_joints=trajectory_dim,
+    model = End2EndDiffusionTransformer(  # TODO enforce all params to be consistent with the dataset
+        num_joints=dataset.num_joints,
         hidden_dim=hidden_dim,
-        num_layers=num_layers,
-        num_heads=num_heads,
-        max_action_context_length=action_context_length,
-        trajectory_prediction_length=trajectory_prediction_length,
+        use_action_history=True,
+        num_action_history_encoder_layers=2,
+        max_action_context_length=100,
+        use_imu=True,
+        imu_orientation_embedding_method=IMUEncoder.OrientationEmbeddingMethod.QUATERNION,
+        num_imu_encoder_layers=2,
+        max_imu_context_length=100,
+        use_joint_states=True,
+        joint_state_encoder_layers=2,
+        max_joint_state_context_length=100,
+        use_images=True,
+        image_sequence_encoder_type=SequenceEncoderType.TRANSFORMER,
+        image_encoder_type=ImageEncoderType.RESNET18,
+        num_image_sequence_encoder_layers=1,
+        max_image_context_length=10,
+        num_decoder_layers=4,
+        trajectory_prediction_length=10,
     ).to(device)
 
     # Add normalization parameters to the model
-    model.mean = torch.tensor(means.values).to(device)
-    model.std = torch.tensor(stds.values).to(device)
+    model.mean = normalizer.mean
+    model.std = normalizer.std
 
+    # Utilize an Exponential Moving Average (EMA) for the model to smooth out the training process
     ema = EMA(model, beta=0.9999)
```

Comment:
For example here I would appreciate a comment, why we choose specifically this beta especially for people like me who have no experience with machine learning.
Rather than comments like:
```python
# Creating test
test = lib.test(...)
```